master/pkg/schemas/defaults.go

```diff
@@ -0,0 +1,233 @@
+package schemas
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// RuntimeDefaultable means there are runtime values for filling in an object, like choosing a
+// random seed based on the wall clock.
+type RuntimeDefaultable interface {
+	// RuntimeDefaults must apply the runtime-defined default values.
+	RuntimeDefaults()
+}
+
+// Defaultable means a struct can have its defaults filled in automatically.
+type Defaultable interface {
+	// DefaultSource must return a parsed json-schema object in which to find defaults.
+	DefaultSource() interface{}
+}
+
+// FillDefaults will recurse through structs, maps, and slices, setting default values for any
+// struct fields whose struct implements the Defaultable interface.  This lets us read default
+// values out of json-schema automatically.
+//
+// There are some forms of defaults which must be filled at runtimes, such as giving a default
+// description to experiments with no description.  This can be accomplished by implementing
+// the RuntimeDefaultable interface for that object.  See ExperimentConfig for an example.
+//
+// There are some objects which get their defaults from other objects' defaults.  This an
+// unforutnate detail of our union types which have common members that appear on the root union
+// object.  That's hard to reason about, and we should avoid doing that in new config objects.  But
+// those objects implement DefaultSource() to customize that behavior.
+//
+// Example usage:
+//
+//    config, err := expconf.ParseAnyExperimentConfigYAML(bytes)
+//
+//    // Use the cluster checkpoint storage if the user did not specify one.
+//    schemas.Merge(&config.CheckpointStorage, cluster_default_checkpoint_storage)
+//
+//    // Define any remaining undefined values.
+//    schemas.FillDefaults(&config)
+//
+func FillDefaults(x interface{}) {
+	obj := reflect.ValueOf(x)
+	if obj.Kind() != reflect.Ptr {
+		panic("FillDefaults must be called on a pointer")
+	}
+	// Enter the recursive default filling with no default bytes for the root object (which must
+	// already exist), and starting with the name of the object type.
+	name := fmt.Sprintf("%T", x)
+	fillDefaults(obj, nil, false, name)
+}
+
+func fillDefaults(obj reflect.Value, defaultBytes []byte, inPtr bool, name string) {
```

Comment:
Seems like there is a general function to recurse `obj` hiding in both the go and python (something something visitor pattern). I know we talked about this a bit but it really may help readability / separate bugs in the recursing from bugs in the defaulting or merging.

Reply:
I actually don't think the recursion is general enough.  FillDefaults is very generic (visit everything and try to fill defaults) but merge is not.

In fact, I realized there were two totally different paths to the recursion of merge, one was intended to behave like a deep copy, and one which did merging.  I think there was potential for bugs with how I had it, so now I have those as two separate recursion paths (and I elimiated the boolean flag that effectively controlled which path you were on).